Add option for ExpansionTile to maintain the state of its children when collapsed

eospi opened the pull request:
## Description
Currently, `ExpansionTile` rebuilds its children upon every expansion and removes the children when collapsed, resetting any state contained in the children. When building a Flutter app a couple months ago, I encountered this issue when placing a `TextFormField` inside an ExpansionTile, and any edited text would be deleted when the user collapsed the ExpansionTile. #37520 further describes the specific issue and includes sample code to reproduce it. I made this change to the `ExpansionTile` widget to make my app work correctly.
### Changes overview:
- Added a `maintainsState` property to `ExpansionTile` of type `bool`
- Set `maintainsState` default value to `false` for the existing behavior
- Add `!maintainsState` as a condition for setting the child of the `AnimatedBuilder` to `null`

Now, an `ExpansionTile` like this maintains the state of its children while collapsed:
```dart
ExpansionTile(
  maintainsState: true,
  key: PageStorageKey("test"),
  title: Text("Click to expand"),
  children: [
    TextFormField(
      key: PageStorageKey("test2"),
      validator: (value) {
        if (value.isEmpty) {
          return 'Please enter some text';
        }
        return null;
      },
    ),
  ]
)
```

## Related Issues

#37520

## Tests

I added the following tests:
- 'ExpansionTile maintainsState' - creates two collapsed ExpansionTile widgets with different `maintainsState` values and tests if a child widget is present in each.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

Commit:
Add maintainsState property to ExpansionTile

Commit:
Use maintainsState to decide ExpansionTile children

Commit:
update documentation

Commit:
set default maintainsState to false

Commit:
simplify boolean logic

Commit:
test maintainsState

Commit:
add const keyword

goderbauer commented on packages/flutter/lib/src/material/expansion_tile.dart:
```diff
@@ -40,6 +40,7 @@ class ExpansionTile extends StatefulWidget {
     this.children = const <Widget>[],
     this.trailing,
     this.initiallyExpanded = false,
+    this.maintainsState = false,
```
nit: in other places of the framework we call a similar setting "maintainState" without the "s". Can you rename this to be consistent?

goderbauer commented on packages/flutter/lib/src/material/expansion_tile.dart:
```diff
@@ -74,6 +75,13 @@ class ExpansionTile extends StatefulWidget {
   /// Specifies if the list tile is initially expanded (true) or collapsed (false, the default).
   final bool initiallyExpanded;
 
+  /// Specifies if the state of the children is maintained when the tile expands and collapses
```
nit: sentences should end with a `.`

goderbauer replied:
```suggestion
  /// Specifies whether the state of the children is maintained when the tile is collapsed.
```

goderbauer commented on packages/flutter/lib/src/material/expansion_tile.dart:
```diff
@@ -197,10 +205,11 @@ class _ExpansionTileState extends State<ExpansionTile> with SingleTickerProvider
       animation: _controller.view,
       builder: _buildChildren,
-      child: closed ? null : Column(children: widget.children),
+      child: shouldRemoveChildren ? null : Column(children: widget.children),
```
Just keeping them in the tree like this doesn't seem like a good idea. Animations will keep running and drain the user's battery even though nothing is visible. We should turn animations off and maybe even put the non-visible children offstage to ensure that you can't interact with them. For an example, look at how the "Visibility" widget implements the maintainState flag.

eospi replied:
Thanks for pointing to the Visibility widget! Should I wrap the children in a Visibility widget or instead wrap it in both Offstage and TickerMode?

goderbauer commented on packages/flutter/test/material/expansion_tile_test.dart:
```diff
@@ -215,4 +215,43 @@ void main() {
+    ));
+
+    // This text should be there while ExpansionTile collapsed
+    expect(find.text('Maintaining State'), findsOneWidget);
```
The tester shouldn't find this if it is not visible unless `skipOffstage: true` is specified.

goderbauer commented:
Can you also rebase these changes to the latest master? That should make (most of) the tests happy, that are currently failing.

eospi commented:
Thanks for looking over the PR! I will begin working on the changes as soon as possible.

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
rename 'maintainsState' to 'maintainState'

Commit:
fix docs

gspencergoog committed:
Revert "Upgrade dartdoc to 0.30.1 (#50980)" (#51060)

This reverts commit a51de38a2064c1c1ed838ce9ec394e31f94386ca because it breaks the flutter docs.

mkustermann committed:
Weaken stack-trace based tests to allow auto-roll of new engine which will change the format (#51047)

Follow-up PR will restore those tests with new expectations:
https://github.com/flutter/flutter/pull/50760

dnfield committed:
manually roll engine to f2f8c342be3243a7a196587e77678e00d7e30c24 (#51071)

miquelbeltran committed:
Add highContrast to FakeAccessibilityFeatures test (#48486)

najeira committed:
Fix comment typo FormField.initialState to FormField.initialValue (#51011)

tjonganthony committed:
[Form] Add is valid to FormState (#48948)

angjieli committed:
Use Async WebDriver for WebFlutterDriver. (#50835)

Commit:
[flutter_tools] move engine unpack to the flutter tool (#50707)

franciscojma86 committed:
Revert "[flutter_tools] move engine unpack to the flutter tool (#50707)" (#51076)

This reverts commit 0a623e879c48f14e7b05b2f4ce377130ab09dfc2.

Commit:
Add 'maintainState' property

Commit:
Add test for 'maintainState'

Commit:
Merge branch 'master' of https://github.com/eospi/flutter

goderbauer commented:
Hello @eospi,

it looks like maybe a merge or rebase has gone wrong in this PR. It now includes a lot of seemingly unrelated changes. Can you fix this please before I take another look at the code? Also, the analyzer is reporting issues with trailing white space: https://github.com/flutter/flutter/pull/41886/checks?check_run_id=456750789

Thanks!

eospi commented:
Is it alright if I open a new PR to fix the extra commit issue?

goderbauer commented:
Go for it! Please close this one when you do.

eospi commented:
I've just opened #51709 as the replacement PR. Thanks!
